lib/rex/proto/ms_dtyp.rb

```diff
@@ -0,0 +1,334 @@
+# -*- coding: binary -*-
+
+module Rex::Proto::MsDtyp
+    # [2.4.2.2 SID--Packet Representation](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/f992ad60-0fe4-4b87-9fed-beb478836861)
+    class MsDtypSid < BinData::Record
+      endian :little
+  
+      uint8 :revision, initial_value: 1
+      uint8 :sub_authority_count, initial_value: -> { sub_authority.length }
+      array :identifier_authority, type: :uint8, initial_length: 6
+      array :sub_authority, type: :uint32, initial_length: :sub_authority_count
+  
+      def assign(val)
+        # allow assignment from the human-readable string representation
+        if val.is_a?(String) && val =~ /^S-1-(\d+)(-\d+)+$/
+          _, _, ia, sa = val.split('-', 4)
+          val = {}
+          val[:identifier_authority] = [ia.to_i].pack('Q>')[2..].bytes
+          val[:sub_authority] = sa.split('-').map(&:to_i)
+        end
+  
+        super
+      end
+  
+      def to_s
+        str = 'S-1'
+        str << "-#{("\x00\x00" + identifier_authority.to_binary_s).unpack1('Q>')}"
+        str << '-' + sub_authority.map(&:to_s).join('-')
+      end
+    end
+
+    # [Universal Unique Identifier](http://pubs.opengroup.org/onlinepubs/9629399/apdxa.htm)
+    # The online documentation at [2.3.4.2](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/001eec5a-7f8b-4293-9e21-ca349392db40)
+    # wierdly doesn't mention this needs to be 4 byte aligned for us to read it correctly,
+    # which the RubySMB::Dcerpc::Uuid definition takes care of.
+    class MsDtypGuid < RubySMB::Dcerpc::Uuid
+    end
+  
+    # [2.4.4.1 ACE_HEADER](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/628ebb1d-c509-4ea0-a10f-77ef97ca4586)
+    class MsDtypAceHeader < BinData::Record
+      endian :little
+  
+      uint8  :ace_type
+      struct :ace_flags do
+        bit1 :failed_access_ace_flag
+        bit1 :successful_access_ace_flag
+        bit1 :reserved
+        bit1 :inherited_ace
+        bit1 :inherit_only_ace
+        bit1 :no_propagate_inherit_ace
+        bit1 :container_inherit_ace
+        bit1 :object_inherit_ace
+      end
+      uint16 :ace_size, initial_value: -> { parent&.num_bytes || 0 }
+    end
+  
+    # [2.4.4.2 ACCESS_ALLOWED_ACE](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/72e7c7ea-bc02-4c74-a619-818a16bf6adb)
+    class MsDtypAccessAllowedAceBody < BinData::Record
+      endian :little
+  
+      uint32             :access_mask
+      ms_dtyp_sid        :sid, byte_align: 4
+    end
+  
+    # [2.4.4.2 ACCESS_ALLOWED_ACE](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/72e7c7ea-bc02-4c74-a619-818a16bf6adb)
+    class MsDtypAccessAllowedAce < BinData::Record
+      endian :little
+  
+      ms_dtyp_ace_header              :header, initial_value: { ace_type: 0 }
+      ms_dtyp_access_allowed_ace_body :body
+    end
```

Comment:
Also, I noticed the ACE structures defined in the specs under https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/d06e5a81-176e-46c6-9cf7-9137aad4455e are very similar and it can be reduced to just a few structures. I'm wondering if it would be possible to avoid code duplication by using inheritance between structures. Something along these lines (probably with a more appropriate class name):

```ruby
    class MsDtypAce < BinData::Record
       endian :little
       ms_dtyp_ace_header :header
    end

    class MsDtypAceType1 < MsDtypAce
       endian :little
       uint32             :access_mask
       ms_dtyp_sid        :sid, byte_align: 4
    end

    class MsDtypAccessAllowedAce < MsDtypAceType1
    end

    class MsDtypAccessDeniedAceBody < MsDtypAceType1
    end
```

Reply:
Not every single entry has an access_mask followed by a SID. Take ACCESS_ALLOWED_OBJECT_ACE which has an access mask, then a flags field, then a object_type field, then a inherited_object_type field, and finally the SID.

This would invalidate the proposed solution since that wouldn't fit all requirements.

I *may* be able to make it so that the `access_mask` gets moved since that seems to always be the first field that is provided but not sure about SID.

Reply:
Looks like all of them at least have a `mask` field directly after the header so this would be possible optimization.

Reply:
https://github.com/rapid7/metasploit-framework/pull/17122/commits/2515183c45fc23afd70f5213b85c5c8d8f1a2401 should now contain the fix for this issue. Thanks for pointing this out to me 👍